File: src/Day03.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashSet;

public class Day03 {

	public static int partOne(String s) {
		return solve(s, 1);
	}

	public static int partTwo(String s) {
		return solve(s, 2);
	}

	public static int solve(String s, int santaCount) {
		HashSet<String> houses = new HashSet<>();
		int[][] pos = new int[santaCount][2];
		houses.add("0:0");
		for (int i = 0; i < s.length(); i++) {
			switch (s.charAt(i)) {
			case '>':
				pos[i % santaCount][0]++;
				break;
			case '^':
				pos[i % santaCount][1]++;
				break;
			case '<':
				pos[i % santaCount][0]--;
				break;
			case 'v':
				pos[i % santaCount][1]--;
				break;
			default:
				break;
			}
			houses.add(pos[i % santaCount][0] + ":" + pos[i % santaCount][1]);
		}
		return houses.size();
	}

	public static void main(String[] args) throws IOException {
		String s = new String(Files.readAllBytes(Paths.get("./input/Day03_input.txt")));
		System.out.println("Part One = " + partOne(s));
		System.out.println("Part Two = " + partTwo(s));
	}
}

File: src/aoc2015/kot/Day03.kt
package aoc2015.kot

import java.io.File
import java.util.*

object Day03 {
  fun partOne(input: String): Set<Pair<Int, Int>> {
    val locations = HashSet<Pair<Int, Int>>()
    var location = Pair(0, 0)
    locations.add(location)
    for (c in input) {
      when (c) {
        '^' -> location = Pair(location.first - 1, location.second)
        'v' -> location = Pair(location.first + 1, location.second)
        '<' -> location = Pair(location.first, location.second - 1)
        '>' -> location = Pair(location.first, location.second + 1)
      }
      locations.add(location)
    }
    return locations
  }
}

fun main(args: Array<String>) {
  val input = File("./input/2015/Day03_input.txt").readText()
  println("Part One = ${Day03.partOne(input).size}")
  val santa = Day03.partOne(input.filterIndexed { idx, _ -> idx % 2 == 0 })
  val santa2 = Day03.partOne(input.filterIndexed { idx, _ -> idx % 2 == 1 })
  println("Part Two = ${santa.union(santa2).size}")
}
